Simplify State/AccountsStore lifecycle by avoiding replacing accounts_db

jasonz-dfinity opened the pull request:
# Motivation

The `AccountsDb::Map(AccountsDbAsMap)` enum value (i.e. representing accounts as a map in heap memory) is deprecated, and we'd like to remove this case (and only have the branch of code for the `AccountsDb::UnboundedStableBTreeMap(AccountsDbAsUnboundedStableBTreeMap<>)`). However, currently it's impossible, since `AccountsStore::decode` calls `AccountsDbAsProxy::default()`, which is an instance of the deprecated type. The `accounts_db` is later [replaced](https://sourcegraph.com/github.com/dfinity/nns-dapp@395ad2a74c9309b2f6b943bce3964921904d6e36/-/blob/rs/backend/src/state.rs?L147)  so that the production code indeed uses the `StableBTreeMap`. Therefore, this PR aims at removing the default-then-replace approach, by initializing the `AccountsDb::UnboundedStableBTreeMap` directly.


# Changes

- Inline and remove `State::save_heap_to_managed_memory` and `State::recover_heap_from_managed_memory`, since the logic is simple (so the indirection is no longer necessary)
- Move some of the logic from `State::new_restored` into `State::decode` and remove the `default`/`replace_accounts_db` logic (which is the main goal of the PR)
- Move the `accounts_store` initialization logic from `State::new` to `AccountsStore::new`, to keep StableBTreeMap initialization logic close to each other

# Tests

N/A since there are no functionality changes. The test in `rs/backend/src/state/tests.rs` should (still) cover the code being changed in this PR.

# Related PRs

[Previous](https://github.com/dfinity/nns-dapp/pull/6301) | [Next](https://github.com/dfinity/nns-dapp/pull/6311)

jasonz-dfinity commented on rs/backend/src/state/tests.rs:
```diff
@@ -31,6 +33,21 @@ fn state_can_be_saved_and_recovered_from_stable_memory(num_accounts: u64) {
+        tvl_state,
+        performance: _, // This field is not persisted through upgrades.
+    } = &state;
+    let expected_accounts: BTreeMap<_, _> = accounts_store.range(..).collect();
+    let mut expected_account_stats = Stats::default();
+    accounts_store.get_stats(&mut expected_account_stats);
+    let expected_assets = assets.clone();
+    let expected_tvl_state = tvl_state.clone();
+    let expected_asset_hashes = asset_hashes.clone();
```
Note: this is a side effect of doing a conversion from `State` to `SerializableState`, since during the conversion, `State` is destroyed, so we need to clone parts of the state before consuming it. An alternative is `impl From<&State> for SerializableState`. However, this would worsen the production code, since the conversion would have to clone data unnecessarily. The downside of having to clone the parts of the state for comparison in a unit test seems to acceptable.

dskloetd commented:
I don't understand what problem you're trying to solve. Can you explain it in more detail?

jasonz-dfinity commented:
> I don't understand what problem you're trying to solve. Can you explain it in more detail?

I read it again myself and I felt some of the indirections were not that necessary. I removed some of the changes and rewrote the PR description. PTAL?

dskloetd reviewed (approved):
Thanks!

dskloetd commented on rs/backend/src/state.rs:
```diff
@@ -1,18 +1,16 @@
 use crate::tvl::state::TvlState;
 
 use dfn_candid::Candid;
+use dfn_core::api::trap_with;
```
Actually, let me just merge this, so I can resolve the conflict already.
